### ISSUE TITLE
@bazel_tools//tools/zip:zipper: allow glob exclusion to prevent files from being extracted from a zip file

### ISSUE THREAD

**Issue by bendavini** [bazelbuild/bazel#23662]
### Description of the feature request:

`unzip` (https://infozip.sourceforge.net/) allows users to specify files to exclude while unzipping a file:

`unzip file.zip -x $(<excl_file.txt)`

In order to use `zipper x` in a similar way, a user needs to perform some additional shell magic to get the same result (we've a full (currently non-hermetic) `unzip.bzl` that uses system unzip to achieve this.)

### Which category does this issue belong to?

CLI, Core

### What underlying problem are you trying to solve with this feature?

Make `zipper`'s extract (`x`) functionally equivalent to the widely-used `unzip` utility

### Which operating system are you running Bazel on?

Mac(x86/aarch64) / Linux (x86)

### What is the output of `bazel info release`?

7.2.1

### Have you found anything relevant by searching the web?

No. Have scoured slack, github, email threads, and our internal tools

**Comment by bendavini** [comment]
TL;DR we want to be able to extract (a subset of) files from a zip archive in a way that is consumable by other actions (mostly `rules_pkg`).

I asked [here](https://bazelbuild.slack.com/archives/CA31HN1T3/p1725387281446829) what is the most proper way to **hermetically** unzip things; the [answer was to use zipper](https://bazelbuild.slack.com/archives/CA31HN1T3/p1725387538986389?thread_ts=1725387281.446829&cid=CA31HN1T3) .

The use case is described in the first link. We have a custom rule which is a wrapper around a non-hermetic `unzip` command that provides an alternative to a `genrule` that does not require the list of files to be extracted.

If `@bazel_tools//tools/zip:zipper` is not meant to be a general purpose zip utility, do you know of something that is (and is also hermetic)?

**Comment by fmeum** [comment]
https://registry.bazel.build/modules/ape has cross-platform binaries for common utilities, including `unzip`.

**Comment by bendavini** [comment]
Thanks @fmeum ! Learn about a new module every day :)

I am happy to use a new module, but would prefer to lean on a built-in if possible. If `zipper` is meant to be exclusive to bazel internals, should I file another ticket to make it be 'advertised' as such? (i.e., something like a readme recommending the APE module for cases like this, reduced visibility of `zipper`, output of said information in the `help`, etc.

**Comment by tjgq** [comment]
@lberki @meteorcloudy Do we have a stance on expanding the scope of `@bazel_tools//tools/zip:zipper` and/or intentionally supporting its use outside of built-in rules?

**Comment by lberki** [comment]
@tjgq @bendavini @ahumesky  I'd much, much rather no one use that built-in zipper tool. It was added in support of Android rules, which are in the process of being removed from the Bazel proper and thus ideally, this binary would also vanish from `@bazel_tools`.

**Comment by haxorz** [comment]
`team-Core` has no involvement or expertise with `@bazel_tools//tools/zip:zipper`

**Comment by meteorcloudy** [comment]
> Do we have a stance on expanding the scope of @bazel_tools//tools/zip:zipper and/or intentionally supporting its use outside of built-in rules?

Definitely not, we want to shrink the size of `@bazel_tools` as much as possible. One day, we may want to even remove it from Bazel builtins.

**Comment by tjgq** [comment]
Since our position is that we don't want to expand scope for `@bazel_tools//tools/zip:zipper` (and possibly even get rid of it in the future), let's close this. Rulesets needing zip functionality should ship their own tool or import it from elsewhere.
